sfe/sfe.go

```diff
@@ -0,0 +1,287 @@
+	if incomingJWT != "" {
+		type tmplData struct {
+			UnpausePath string
+			JWT         string
+		}
+
+		// Serve the actual unpause page given to a Subscriber. Populates the
+		// unpause form with the JWT from the URL. That JWT itself may be
+		// invalid or expired, but that validation will be performed only after
+		// submitting the form.
+		renderTemplate(response, "unpause-params.html", tmplData{unpausePath, string(incomingJWT)})
+	} else {
+		// We only want to accept requests containing the JWT param.
+		renderTemplate(response, "unpause-noParams.html", nil)
+	}
+}
+
+// posthelper After clicking unpause, serve a page indicating if the unpause succeeded or failed.
+func (sfe *SelfServiceFrontEndImpl) postHelper(response http.ResponseWriter, incomingJWT unpauseJWT) {
+	if incomingJWT != "" {
+		type tmplData struct {
+			ShouldUnpause bool
+			AccountID     string
+		}
+
+		claims, err := sfe.validateJWTforAccount(incomingJWT)
+		if err != nil {
+			renderTemplate(response, "unpause-post.html", nil)
+		}
+
+		// TODO(#7356) Declare a registration ID variable to populate an
+		// rapb unpause account request message.
+		_, innerErr := strconv.ParseInt(claims.Subject, 10, 64)
+		if innerErr != nil {
+			renderTemplate(response, "unpause-noParams.html", nil)
+		}
+
+		// TODO(#7536) Send gRPC request to the RA informing it to unpause
+		// the account specified in the claim. At this point we should wait
+		// for the RA to process the request before returning to the client,
+		// just in case the request fails.
+
+		// Success, the account has been unpaused.
+		renderTemplate(response, "unpause-post.html", tmplData{true, claims.Subject})
+	} else {
+		renderTemplate(response, "unpause-noParams.html", nil)
+	}
+}
+
+// Unpause allows a requester to unpause their account via a form present on the
+// page.
+func (sfe *SelfServiceFrontEndImpl) Unpause(response http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodHead:
+		return
+	case http.MethodGet:
+		sfe.getHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	case http.MethodPost:
+		sfe.postHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	default:
+		response.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, POST")
+		response.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+}
+
+type sfeJWTClaims struct {
+	jwt.Claims
+
+	// Version is a custom claim used to mass invalidate existing JWTs by
+	// changing the API version via unpausePath.
+	Version string `json:"apiVersion,omitempty"`
+}
+
+// validateJWT derives a ed25519 public key from a seed shared by the SFE and
+// WFE. The public key is used to validate the signature and contents of an
+// unpauseJWT and verify that the its claims match a set of expected claims.
+// Passing validations returns the claims or an error.
+func (sfe *SelfServiceFrontEndImpl) validateJWTforAccount(incomingJWT unpauseJWT) (sfeJWTClaims, error) {
```

Comment:
This doesn't seem to validate that a JWT is valid for a given account (ideally that of the requester), only that it was issued by a similarly configured WFE.

Reply:
Since the WFE creates the JWT, it's the WFEs job to determine what the to-be-unpaused account ID is. The SFE has no knowledge of what/who the account belongs too. It will simply route the account ID back to the RA via gRPC for actual unpausing (assuming that the JWT was able to be validated).

Given that the WFE and SFE share a seed value for signing/validating a JWT, we are able to trust that the account ID wasn't tampered in transit from WFE to SFE.

Reply:
I'm only saying that this doesn't do what it says it does. It doesn't validate a JWT for an account, it validates an unpause JWT.